### ISSUE TITLE
Repetition penalty + sampling controls — model degeneration in autoregressive loop

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/ATLAS#16]
## Summary

`generate()` produces repetitive output on OLMo-3-7B-Think even with correct EOS/PRNG/ChatML fixes (v4.0.5). The model assigns ~99.9% probability to a single token once it enters a repetition loop, making sampling ineffective. Standard inference-time controls (repetition penalty, top-p, top-k) are missing.

## Evidence

GPU E2E test on A100-SXM4-40GB, OLMo-3-7B-Think (BF16, 14.6GB):

```
Prompt: <|im_start|>user\nWhat is the capital of France?<|im_end|>\n<|im_start|>assistant\n

Greedy (temp=0.0):
  " subsid subsid subsid subsid subsid subsid subsid subsid..." (30× token 32711)

Temp=1.0 sampling (proper XorShift64 PRNG):
  " subsid subsid subsid subsid subsid subsid subsid subsid..." (identical)

Temp=0.8 sampling:
  "caffecaffecaffecaffecaffecaffecaffe..." (30× same token)
```

The model's softmax distribution is so peaked (~99.9% on one token) that even temperature=1.0 sampling picks the same token every step. Two independent calls with different PRNG seeds produce identical output — not because the PRNG is broken (Issue #14, fixed), but because the probability mass is overwhelmingly concentrated.

### Why the old code looked "diverse"

Before v4.0.5, `sample_from_probs(step as u64)` used a deterministic LCG hash of the step number. This produced a fixed sequence of "random" values (0.078, 0.965, 0.259, ...) that happened to land in the 0.1% tail of the distribution at some steps, selecting different tokens. The output *looked* varied but was actually:
- Deterministic (same output every call)
- Incoherent (random tokens, not meaningful text)
- Masking the real problem (model degeneration)

## Root Cause

Autoregressive repetition loops are a well-known transformer failure mode:
1. Model predicts token X with high confidence
2. Token X is added to context
3. Context with X reinforces prediction of X (self-amplifying)
4. After 2–3 repetitions, probability of X approaches 1.0
5. No amount of random sampling can escape — the distribution is too peaked

This is **not a model quality issue** — it happens with GPT-2, LLaMA, and virtually all autoregressive LLMs without inference-time controls. Every production inference engine (vLLM, TGI, llama.cpp, OpenAI API) implements these controls.

### 1. Repetition Penalty (critical)

Divide logits of previously-seen tokens by a penalty factor before softmax.

```rust
// Before softmax, for each token in the generated history:
for &prev_token in &new_tokens {
    if logits[prev_token as usize] > 0.0 {
        logits[prev_token as usize] /= repetition_penalty; // e.g. 1.1–1.3
    } else {
        logits[prev_token as usize] *= repetition_penalty;
    }
}
```

OpenAI API parameter: not directly exposed, but used internally.
HuggingFace `transformers` parameter: `repetition_penalty` (default 1.0 = off).

### 2. Top-p / Nucleus Sampling (high priority)

After temperature scaling + softmax, sort by probability descending. Keep only the smallest set of tokens whose cumulative probability ≥ p. Renormalize and sample.

```rust
fn top_p_filter(probs: &mut Vec<(usize, f32)>, p: f32) {
    probs.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
    let mut cum = 0.0;
    let mut cutoff = probs.len();
    for (i, &(_, prob)) in probs.iter().enumerate() {
        cum += prob;
        if cum >= p {
            cutoff = i + 1;
            break;
        }
    }
    probs.truncate(cutoff);
    // Renormalize
    let sum: f32 = probs.iter().map(|(_, p)| p).sum();
    for (_, p) in probs.iter_mut() { *p /= sum; }
}
```

OpenAI API parameter: `top_p` (default 1.0 = off).

### 3. Top-k Sampling (nice to have)

Keep only the top-k highest-probability tokens. Simpler than top-p.

OpenAI API parameter: not directly exposed.
Common in llama.cpp, HuggingFace: `top_k` (default 50).

### 4. Frequency / Presence Penalty (nice to have)

OpenAI-style penalties applied to logits before softmax:
- **frequency_penalty**: subtract `penalty * count(token)` from logit (penalizes frequent tokens)
- **presence_penalty**: subtract `penalty * (count > 0 ? 1 : 0)` from logit (penalizes any repeated token)

```rust
for (token, count) in token_counts.iter() {
    logits[*token] -= frequency_penalty * (*count as f32);
    if *count > 0 {
        logits[*token] -= presence_penalty;
    }
}
```

OpenAI API parameters: `frequency_penalty` (default 0.0), `presence_penalty` (default 0.0).

### Phase 1 (this issue): `generate()` internal controls

Add to `OlmoModel::generate()`:

```rust
pub fn generate(
    &mut self,
    prompt: &[u32],
    max_new: usize,
    temperature: f32,
    repetition_penalty: f32,  // 1.0 = off, 1.1–1.3 typical
    top_p: f32,               // 1.0 = off, 0.9–0.95 typical
    top_k: usize,             // 0 = off, 40–100 typical
) -> Vec<u32>
```

Or use a config struct:

```rust
pub struct SamplingParams {
    pub temperature: f32,        // default 1.0
    pub repetition_penalty: f32, // default 1.0 (off)
    pub top_p: f32,              // default 1.0 (off)
    pub top_k: usize,            // default 0 (off)
    pub frequency_penalty: f32,  // default 0.0
    pub presence_penalty: f32,   // default 0.0
}
```

### Phase 2: Wire through atlas-api

Parse from OpenAI-compatible request JSON:
```json
{
  "temperature": 0.7,
  "top_p": 0.9,
  "frequency_penalty": 0.5,
  "presence_penalty": 0.3,
  "repetition_penalty": 1.1
}
```

Pass to `run_inference()` → `model.generate()`.

### Phase 3: Defaults that work

Set sensible defaults for chat models:
- `repetition_penalty: 1.1`
- `top_p: 0.9`
- `temperature: 0.7`

These match what most OpenAI-compatible clients expect.

## Complexity Estimate

| Feature | Lines of code | Risk |
|---------|--------------|------|
| Repetition penalty | ~15 in generate() | None |
| Top-p sampling | ~25 (sort + cumsum + truncate) | Low |
| Top-k sampling | ~10 (sort + truncate) | None |
| Frequency/presence penalty | ~15 | None |
| SamplingParams struct | ~20 | None |
| API wiring (parse + pass) | ~30 | Low |
| Tests | ~60 | None |
| **Total** | **~175** | **Low** |

## Environment
- ATLAS v4.0.5 (549 tests)
- Affects all models served via `atlas api serve`
- Unblocks functional chat with OLMo-3-7B-Think
- Required before HuggingFace demo upload

**Comment by web3guru888** [comment]
## Research Report: Sampling Controls Implementation Plan

**Full report**: [repetition-penalty-research-2026-04-18.md](https://github.com/web3guru888/ATLAS/wiki) (38KB, 10 techniques analyzed, Rust pseudocode included)

### P0 — Fix Degeneration (this issue)

**Root cause**: After EOS/PRNG/ChatML fixes (v4.0.5), the model still assigns ~99.9% probability to a single token once it enters a repetitive loop. Three interventions needed, applied in this order:

| # | Component | Default | Effort | Source |
|---|-----------|---------|--------|--------|
| 1 | **Repetition penalty** | θ=1.1, window=64 | 1–2h | Keskar 2019 (CTRL) — universal standard |
| 2 | **Temperature** | T=0.6 | 30min | OLMo-3-7B-Think model card |
| 3 | **Top-P (nucleus)** | p=0.95 | 2–3h | OLMo-3-7B-Think model card, Holtzman 2020 |

**Pipeline**: `logits → rep_penalty → temperature → top_p → sample`

### Implementation sketch

```rust
pub struct SamplingConfig {
    pub repetition_penalty: f32,  // 1.0 = off, 1.1 = default
    pub repetition_window: usize, // 64 tokens lookback
    pub temperature: f32,         // 0.6 for OLMo-3-7B-Think
    pub top_p: f32,               // 0.95 for OLMo-3-7B-Think
    pub top_k: usize,             // 0 = disabled
    pub min_p: f32,               // 0.0 = disabled (future: 0.05)
}
```

**Repetition penalty** (Keskar): For each token in the last `window` tokens, if `logit > 0`, divide by θ; if `logit < 0`, multiply by θ. This asymmetry ensures the penalty always _reduces_ the probability.

### Key findings from research

- **Min-P** (ICLR 2025 oral) is simpler than top-p and produces better results — implement as P1 follow-up
- **Frequency penalty is DANGEROUS** for reasoning models — causes cycling in long traces (Salesforce LZ Penalty paper, April 2025)
- **LZ Penalty** (arXiv:2504.20131) is SOTA — enables greedy decoding without degeneration, good P3 target
- **DRY sampling** catches phrase-level repetition that token penalties miss — good P2 addition

### Roadmap

- **P0**: rep_penalty + temperature + top_p ← **this PR**
- **P1**: min_p + top_k + presence_penalty (~3h)
- **P2**: DRY sampling for n-gram repetition (~6h)
- **P3**: LZ Penalty for SOTA anti-degeneration (~2d)

**Comment by web3guru888** [comment]
## ✅ Issue #16 Resolved — Sampling Controls (v4.0.6-rc)

**Commit**: [`a257126`](https://github.com/web3guru888/ATLAS/commit/a257126)

### What was implemented

Full sampling pipeline added to `atlas-model` and `atlas-api`:

| Control | Parameter | Default | Recommended |
|---------|-----------|---------|-------------|
| Repetition penalty | `repetition_penalty` | 1.0 (off) | 1.1 |
| Frequency penalty | `frequency_penalty` | 0.0 (off) | 0.1 |
| Presence penalty | `presence_penalty` | 0.0 (off) | 0.3 |
| Top-P (nucleus) | `top_p` | 1.0 (off) | 0.95 |
| Top-K | `top_k` | 0 (off) | 50 |
| Min-P | `min_p` | 0.0 (off) | 0.05 |
| Temperature | `temperature` | 1.0 | 0.6 |

**Pipeline order**: rep_penalty → freq_penalty → pres_penalty → temperature → top_k → top_p → min_p → softmax → sample

### Files changed (3 files, +521/-24 lines)
- `crates/atlas-model/src/lib.rs` — `SamplingConfig` struct, 6 sampling functions, `generate_with_sampling()`
- `crates/atlas-api/src/types.rs` — New fields on `ChatCompletionRequest` / `CompletionRequest`
- `crates/atlas-api/src/handler.rs` — `run_inference()` now accepts `SamplingConfig`

### Tests: 562 passed (+13 new), 0 failed
```
repetition_penalty_reduces_repeated ✅
repetition_penalty_one_is_noop ✅
top_p_filters_low_prob ✅
top_p_one_keeps_all ✅
top_k_keeps_only_k ✅
top_k_zero_is_noop ✅
min_p_filters_low_relative ✅
frequency_penalty_proportional ✅
presence_penalty_flat ✅
sampling_config_default_is_passthrough ✅
sampling_config_olmo3 ✅
generate_backward_compat ✅
generate_with_repetition_penalty ✅
```

### API usage
```bash
# OpenAI-compatible API call with sampling controls
curl http://localhost:8080/v1/chat/completions -d '{
  "model": "atlas",
  "messages": [{"role": "user", "content": "Hello"}],
  "temperature": 0.6,
  "top_p": 0.95,
  "repetition_penalty": 1.1,
  "max_tokens": 256
}'
```

### Rust API
```rust
use atlas_model::{SamplingConfig, OlmoModel};

// Use OLMo-3 recommended defaults
let config = SamplingConfig::olmo3(); // temp=0.6, rep=1.1, top_p=0.95
let tokens = model.generate_with_sampling(&prompt, 256, &config);

// Or customize
let config = SamplingConfig {
    temperature: 0.8,
    repetition_penalty: 1.2,
    top_p: 0.9,
    top_k: 50,
    ..SamplingConfig::default()
};
```

Zero external dependencies maintained. Backward-compatible — `generate()` delegates to `generate_with_sampling()` with default config.
